File: DSA/Aanidhay Aggarwal/Day 8/Day8_q2.java
import java.util.LinkedList;
import java.util.Queue;

class MyStack {
    private Queue<Integer> queue1;
    private Queue<Integer> queue2;
    private int topElement;

    public MyStack() {
        queue1 = new LinkedList<>();
        queue2 = new LinkedList<>();
    }

    public void push(int x) {
        queue1.add(x);
        topElement = x;
    }
    public int pop() {

        while (queue1.size() > 1) {
            topElement = queue1.remove();
            queue2.add(topElement);
        }

        int poppedElement = queue1.remove();

        Queue<Integer> temp = queue1;
        queue1 = queue2;
        queue2 = temp;

        return poppedElement;
    }

    public int top() {
        return topElement;
    }
    
    public boolean empty() {
        return queue1.isEmpty();
    }

    public static void main(String[] args) {
        MyStack stack = new MyStack();
        stack.push(1);
        stack.push(2);
        System.out.println(stack.top());
        System.out.println(stack.pop());
        System.out.println(stack.top());
    }
}

File: DSA/SparshJain/Day8_Q2.cpp
#include <iostream>
#include <queue>
using namespace std;

class Stack {
private:
    queue<int> q1, q2;
    
    void transfer(queue<int>& from, queue<int>& to) {
        while (from.size() > 1) {
            to.push(from.front());
            from.pop();
        }
    }

public:
    void push(int x) {
        q1.push(x);
    }

    void pop() {
        if (q1.empty()) return;
        transfer(q1, q2);
        q1.pop();
        swap(q1, q2);
    }

    int top() {
        if (q1.empty()) return -1;
        transfer(q1, q2);
        int top = q1.front();
        q2.push(top);
        q1.pop();
        swap(q1, q2);
        return top;
    }

    bool empty() {
        return q1.empty();
    }
};

int main() {
    Stack stack;
    stack.push(1);
    stack.push(2);
    cout << stack.top() << endl; // Output: 2
    stack.pop();
    cout << stack.top() << endl; // Output: 1
    stack.pop();
    cout << stack.empty() << endl; // Output: 1 (true)
    return 0;
}
